balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,658 @@
+	// pickfirst.
+	Name = "pick_first_leaf"
+)
+
+const logPrefix = "[pick-first-leaf-lb %p] "
+
+type pickfirstBuilder struct{}
+
+func (pickfirstBuilder) Build(cc balancer.ClientConn, _ balancer.BuildOptions) balancer.Balancer {
+	ctx, cancel := context.WithCancel(context.Background())
+	b := &pickfirstBalancer{
+		cc:               cc,
+		addressList:      addressList{},
+		subConns:         resolver.NewAddressMap(),
+		serializer:       grpcsync.NewCallbackSerializer(ctx),
+		serializerCancel: cancel,
+		state:            connectivity.Connecting,
+	}
+	b.logger = internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf(logPrefix, b))
+	return b
+}
+
+func (b pickfirstBuilder) Name() string {
+	return Name
+}
+
+func (pickfirstBuilder) ParseConfig(js json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	var cfg pfConfig
+	if err := json.Unmarshal(js, &cfg); err != nil {
+		return nil, fmt.Errorf("pickfirst: unable to unmarshal LB policy config: %s, error: %v", string(js), err)
+	}
+	return cfg, nil
+}
+
+type pfConfig struct {
+	serviceconfig.LoadBalancingConfig `json:"-"`
+
+	// If set to true, instructs the LB policy to shuffle the order of the list
+	// of endpoints received from the name resolver before attempting to
+	// connect to them.
+	ShuffleAddressList bool `json:"shuffleAddressList"`
+}
+
+// scData keeps track of the current state of the subConn.
+type scData struct {
+	// The following fields are initialized at build time and read-only after
+	// that.
+	subConn balancer.SubConn
+	addr    resolver.Address
+
+	// The following fields should only be accessed from a serializer callback
+	// to ensure synchronization.
+	state   connectivity.State
+	lastErr error
+}
+
+func (b *pickfirstBalancer) newSCData(addr resolver.Address) (*scData, error) {
+	sd := &scData{
+		state: connectivity.Idle,
+		addr:  addr,
+	}
+	sc, err := b.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{
+		StateListener: func(state balancer.SubConnState) {
+			// Store the state and delegate.
+			b.serializer.TrySchedule(func(_ context.Context) {
+				sd.state = state.ConnectivityState
+				b.updateSubConnState(sd, state)
+			})
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	sd.subConn = sc
+	return sd, nil
+}
+
+type pickfirstBalancer struct {
+	// The following fields are initialized at build time and read-only after
+	// that and therefore do not need to be guarded by a mutex.
+	logger *internalgrpclog.PrefixLogger
+	cc     balancer.ClientConn
+
+	// The serializer and its cancel func are initialized at build time, and the
+	// rest of the fields here are only accessed from serializer callbacks (or
+	// from balancer.Balancer methods, which themselves are guaranteed to be
+	// mutually exclusive) and hence do not need to be guarded by a mutex.
+	// The serializer is used to ensure synchronization of updates triggered
+	// from the idle picker and the already serialized resolver,
+	// subconn state updates.
```

Comment:
> Java and c-core make use of their implementations of a callback serializer to manage these, so I decided to do the same.

Java and C _require_ the callback serializer to do anything.  That's not the way we manage synchronization in gRPC-Go, unless there is the potential for re-ordering of operations that would cause problems, or some kind of re-entrancy problem or something.

I think the mutex approach would be better, since it's much simpler to do the synchronous operations (no need to make a channel, schedule, block -- and reason about whether that's valid), and doesn't require spawning & managing a separate goroutine, or passing around closures.